ideascube/serveradmin/catalog.py

```diff
@@ -604,9 +606,19 @@ def list_nothandled(self, ids):
 
         return sorted(pkgs, key=attrgetter('id'))
 
+    def _update_displayed_packages_on_home(self, *, to_remove_ids=None, to_add_ids=None):
```

Comment:
# General answers 

> Only local stuff should use this method, so when calling it we'd have the source to see what the parameters should be. (and it is private, so it's not API, so it's fine to have to read the code to understand it IMHO)

I disagree here. A function is made to encapsulate a implementation. If you have to read the source code to know how to use a function you failed somewhere. Whatever the function is public or private.
I don't say that you don't need to read the code to understand the function in this case (there is no docstring, ...) but I disagree with the argument "it's private, we don't care of the interface".

In `def foo(bar=None): ...` bar is not a keyword argument, it's a argument with a default value.

> It's just too easy to mistake it for a _ and think the function takes a positional argument that it doesn't use.

Then python will complain if you pass a positional argument.

>  It is also so rarely used, and so hard to search for (try searching for "python *" on a search engine :stuck_out_tongue_winking_eye:) that it certainly would cause trouble to a new developer.

Well, if nobody use it cause nobody use it, nobody will use it :wink: 

> Basically what I'm saying is that I dislike this * construction and I think it shouldn't be used because the benefits are negligible compared to the benefits, as long as we are careful to always pass kwargs as kwargs and never as positional args.

Using `*` is the way to define a argument as a kwarg and not as a positional arg.
In python2 there is no way to define a kwarg but use a `**kwargs` and look into the dict (and a a docstring to tell which kwarg you can pass)
The default value is not related to the type of the argument.

I prefer to force the use of kwargs when the arguments are optional and the user will use a combination of arguments depending of its use.
There is no arguments "stronger" than an other. If you use positional argument, you imply an order, you can omit the second while passing the first but not the contrary.

# Our case

`to_remove_ids` is not "superior" to `to_add_ids`. Both are at the same level and one can be use without the other (not with the default value for the other, **without the other**)

With `def _update_displayed_packages_on_home(self, to_remove_ids=None, to_add_ids=None)` you can write :
```python
self._update_displayed_packages_on_home(a_list)
self._update_displayed_packages_on_home(a_list, another_list)
self._update_displayed_packages_on_home(a_list, to_add_ids=another_list)
self._update_displayed_packages_on_home(to_remove_ids=a_list)
self._update_displayed_packages_on_home(to_add_ids=a_list)
self._update_displayed_packages_on_home(to_remove_ids=a_list, to_add_ids=another_list)
```

With `def _update_displayed_packages_on_home(self, *, to_remove_ids=None, to_add_ids=None)` you can write :
```python
self._update_displayed_packages_on_home(to_remove_ids=a_list)
self._update_displayed_packages_on_home(to_add_ids=a_list)
self._update_displayed_packages_on_home(to_remove_ids=a_list, to_add_ids=another_list)
```

I prefer the second version cause, as you mention, it enforces passing kwargs as kwargs (cause you specify that `to_remove_ids` and `to_add_ids` are kwargs)

If you don't what a user to use `foo('value')` cause you think he should use a kwarg, explicit it !
As someone said, "Explicit is better than implicit"  :stuck_out_tongue: